### PR TITLE
Add nrow and ncol fallback implementations

### PR DESCRIPTION
**Pull request by bkamins**
To help with issues like https://github.com/JuliaData/Tables.jl/issues/342

### CODE REVIEW

**Comment by bkamins** [comment]
I can see that my implementation is incorrect, but let us wait for @quinnj to comment how this should be fixed (my implementation does not work correctly for all tables).

**Comment by juliohm** [comment]
Thank you Bogumil for addressing this issue. It would be really nice to get more explicit support in Tables.jl for `nrow` and `ncol`, especially `nrow`.

A few comments:

1. Julia Base has a different naming convention for functions that return integers: `ndims`, `nfields`, `ndigits`, ... all end with "s". Is there a reason for `nrow` and `ncol` to be singular? Would it be too breaking to rename them to `nrows` and `ncols` for consistency?
2. Is there a reason why `nrow` and `ncol` are implemented in DataAPI.jl? Could these names be moved to Tables.jl in place of `Tables.rowcount`? It feels like the most natural home for these functions.
3. I understand that Tables.jl also targets the use case where tables have infinite streams of rows. I think this use case could be considered the exception instead of the default, and that `nrow` and `ncol` were part of the default table trait. Infinite tables could be formalized with a special trait that only a few users would benefit.
4. It is important to emphasize that adding `nrow` to the default table trait doesn't imply that all tables will fit in memory. We can have the a priori knowledge of `nrow` from a file format specification, and still only be able to load chunks of the file in memory because it is too big.

**Comment by bkamins** [comment]
> Is there a reason for `nrow` and `ncol` to be singular?

This is just a convention that was already in DataFrames.jl before I started developing it. The convention originates from R I guess.
So this is kind of a situation where the names were picked because everyone used them. It is like in the English language (I am not a native speaker so my perspective does not have to be super correct): people say "gonna" because it is short and everyone says so, although I was taught that this is technically incorrect.

> It feels like the most natural home for these functions.

They could. The reason is that Tables.jl provides implementations while DataAPI.jl provides only API. And as such `nrow` and `ncol` do not have currently a default implementation (they could get it in Tables.jl).

> I understand that Tables.jl also targets the use case where tables have infinite streams of rows.

Yes. That is why I have commented above that what should `nrow`/`ncol` return in case of a lazy table that is e.g. schemaless and of unknown number of rows would have to be decided as a part of the design.

> It is important to emphasize that adding `nrow` to the default table trait doesn't imply that all tables will fit in memory.

Yes. `nrow` should play with `Tables.subset` (so that users know what indices are valid)

**Comment by bkamins** [comment]
@quinnj - I have updated the PR following your comment on `rowcount` usage. Can you please have a look if its design is OK now? If yes I will add tests.

File: docs/src/index.md [hunk 1]
```diff
@@ -229,19 +229,21 @@ make my custom type valid for Tables.jl consumers?
...
-| `Tables.columns(x::MyTable)`                |                              | Return an `Tables.AbstractColumns`-compatible object from your table                                                            | 
-| **Optional methods**                   |                              |                                                                                                                                 |
-| `Tables.schema(x::MyTable)`                     | `Tables.schema(x) = nothing` | Return a [`Tables.Schema`](@ref) object from your `Tables.AbstractRow` iterator or `Tables.AbstractColumns` object; or `nothing` for unknown schema |
-| `Tables.materializer(::Type{MyTable})` | `Tables.columntable`         | Declare a "materializer" sink function for your table type that can construct an instance of your type from any Tables.jl input |
-| `Tables.subset(x::MyTable, inds; viewhint)`  |            | Return a row or a sub-table of the original table
+| Required Methods                             | Default Definition           | Brief Description                                                                                                                                   |
```
Comment (bkamins) [inline docs/src/index.md:232]: Here I only added `nrow` and `ncol` lines (at the end). The rest is just the whitespace alignment of a table.

**Review by quinnj — APPROVED** [review]
@bkamins, this LGTM; would you mind adding some simple tests that cover the new `nrow`/`ncol` definitions for the various Tables-owned table types? With that, feel free to merge and we can tag a new release.

**Comment by bkamins** [comment]
Thank you!

**Pull request merged**